src/contracts/MultiLP.sol

```diff
@@ -0,0 +1,250 @@
+
+    constructor(address _liquidityAsset) {
+        require(_liquidityAsset == address(token0) || _liquidityAsset == address(token1), "invalid liquidity asset");
+        liquidityAsset = _liquidityAsset;
+    }
+
+    /// @dev called by the concrete contract's `initialize` function
+    function _initMultiLP(string calldata _name, string calldata _symbol) internal {
+        __ERC20_init(_name, _symbol);
+
+        // Transfer a small bit of liquidity from the intializer to this contract
+        IERC20(liquidityAsset).transferFrom(msg.sender, address(this), MIN_TOTAL_SUPPLY);
+
+        // mint a small amount of shares to a dead account so the total supply can never be zero
+        // This avoids donation attacks when there are no assets in the ARM contract
+        _mint(DEAD_ACCOUNT, MIN_TOTAL_SUPPLY);
+    }
+
+    /// @notice Preview the amount of shares that would be minted for a given amount of assets
+    /// @param assets The amount of liquidity assets to deposit
+    /// @return shares The amount of shares that would be minted
+    function previewDeposit(uint256 assets) public view returns (uint256 shares) {
+        shares = convertToShares(assets);
+    }
+
+    /// @notice deposit liquidity assets in exchange for liquidity provider (LP) shares.
+    /// The caller needs to have approved the contract to transfer the assets.
+    /// @param assets The amount of liquidity assets to deposit
+    /// @return shares The amount of shares that were minted
+    function deposit(uint256 assets) external returns (uint256 shares) {
+        _preDepositHook();
+
+        shares = convertToShares(assets);
+
+        // Transfer the liquidity asset from the sender to this contract
+        IERC20(liquidityAsset).transferFrom(msg.sender, address(this), assets);
+
+        // mint shares
+        _mint(msg.sender, shares);
+
+        _postDepositHook(assets);
+    }
+
+    function _preDepositHook() internal virtual;
+    function _postDepositHook(uint256 assets) internal virtual;
+
+    /// @notice Preview the amount of assets that would be received for burning a given amount of shares
+    /// @param shares The amount of shares to burn
+    /// @return assets The amount of liquidity assets that would be received
+    function previewRedeem(uint256 shares) public view returns (uint256 assets) {
+        assets = convertToAssets(shares);
+    }
+
+    /// @notice Request to redeem liquidity provider shares for liquidity assets
+    /// @param shares The amount of shares the redeemer wants to burn for liquidity assets
+    /// @return requestId The index of the withdrawal request
+    /// @return assets The amount of liquidity assets that will be claimable by the redeemer
+    function requestRedeem(uint256 shares) external returns (uint256 requestId, uint256 assets) {
+        _preWithdrawHook();
+
+        // burn redeemer's shares
+        _burn(msg.sender, shares);
```

Comment:
I think there's a mistake here, by burning the shares before calculating the amount of assets to transfer.

Burning the shares before calculating the amount of assets to transfer to the user results in an under-estimated value in total supply. This leads to an over-estimated amount of assets to transfer to the user. 
PoC with the failing test here: https://github.com/OriginProtocol/arm-oeth/blob/b2c98cbdec2a6a72cce408f6af616b3e54f2b92f/test/fork/LidoFixedPriceMultiLpARM/RequestRedeem.t.sol#L53 

I think doing the opposite might solve the issue:
``` solidity
...
// Calculate the amount of assets to transfer to the redeemer
assets = previewRedeem(shares);

 // burn redeemer's shares
_burn(msg.sender, shares);
...
```
It is done like this in OZ ERC4626 implementation: 
https://github.com/OpenZeppelin/openzeppelin-contracts/blob/e50c24f5839db17f46991478384bfda14acfb830/contracts/token/ERC20/extensions/ERC4626.sol#L187

Reply:
you are correct. I'll move the burn further down the `requestRedeem` function. Thanks